## AfterQC

- Different from most tools, AfterQC analyses the overlapping of paired sequences for pair-end sequencing data. Based on overlapping analysis, AfterQC can detect and cut adapters, and furthermore it gives a novel function to correct wrong bases in the overlapping regions. 

- filtersoutbadreads,detectsandeliminatessequencerâ€™sbubble effects, trims reads at front and tail, detects the sequencing errors and corrects part of them, and finally outputs clean data and generates HTML reports with interactive figures

  

- Bubble detection and visualisationï¼šä½¿ç”¨èšç±»çš„æ–¹å¼æ£€æµ‹å‡ºæ°”æ³¡åœ¨å“ªå¹¶ä¸”è¿›è¡Œå¯è§†åŒ–å±•ç¤º

- Automatic trimmingï¼šå³ä¿®å»ºå¼€å¤´å’Œç»“å°¾çš„ä½è´¨é‡ç¢±åŸºï¼Œä¸ºäº†é¿å…å±€éƒ¨ç­–ç•¥çš„å¼Šç«¯ï¼Œæå‡ºäº†ä¸€ç§å…¨å±€ç­–ç•¥ï¼šå…ˆåšä¸€éæ•´ä½“çš„ç»Ÿè®¡ï¼Œç”»å‡ºè´¨é‡æ›²çº¿ï¼›ç„¶åä»ä¸­é—´å¼€å§‹å¾€ä¸¤è¾¹æ‰«ç›´åˆ°é‡åˆ°ä¸æ­£å¸¸çš„éƒ¨åˆ†(  

  

  1), too high or too low of mean base content percentages (i.e higher than 40%, or lower than 15%); 

  2), too significant change of mean base content percentages (i.e, Â±10% change com- paring to neighbour cycle); 

  3), too high or too low of mean GC percentages (i.e higher than 70%, or lower than 30%); 

  4), too low of mean quality (i.e. less then Q20). Figure 4 gives an example how automatic trimming works.

  

     )->Before trimming happens, AfterQC will do pre-filtering quality control to calculate the base content and quality curves. Our algorithm initialises the central cycle as a good cycle, and then expands the good region by scanning the base content and quality curves cycle by cycle, until it meets the front or end, or meet a cycle con- sidered as abnormal. Then the cycles in the good region will be kept, and the rest cycles in the front and tail will be trimmed. 

- Filteringï¼šquality filters->count the number of low quality bases or N, calculate the mean quality of each read, and then determine whether to keep or discard this read.   polyX filters://TODO

- Overlapping analysis and error correctionï¼šæšä¸¾offestæ‰¾åˆ°åŒç«¯æ•°æ®çš„é‡å éƒ¨åˆ†ï¼ˆè®¡ç®—æœ€å°çš„ç¼–è¾‘è·ç¦»//TODOï¼‰ï¼Œç„¶åè¿›è¡Œä¿®æ­£->å¦‚æœæµ·æ˜è·ç¦»==ç¼–è¾‘è·ç¦»ï¼Œå°±ç›´æ¥ä¿®æ­£æ²¡å¯¹ä¸ä¸€æ ·çš„ç¢±åŸºï¼ˆè´¨é‡åˆ†ä½çš„æ”¹æˆé«˜çš„ï¼‰//TODO

- Sequencing error profilingï¼šå³æ ¹æ®ğŸ‘†æåˆ°çš„ä½è´¨é‡æ”¹ä¸ºé«˜è´¨é‡åšä¸‹ç»Ÿè®¡ï¼Œç„¶ååˆ†æä¸€ä¸‹ï¼Œå‘ç°è¿™ä¸ªä¾èµ–äºæµ‹åºä»ªã€‚

- Automatic adapter cuttingï¼šè¿™é‡Œåªè®²åˆ°äº†åŒç«¯æ•°æ®å»é™¤adapterï¼Œæ–¹å¼å°±æ˜¯åŸºäºoverloaping analyzeå¦‚æœæ‰¾åˆ°çš„best offest < 0ï¼Œå°±æ–­å®šå‡ºç°äº†ä¸‹å›¾çš„çŠ¶å†µï¼š![image-20210701102351548](/Users/ylf9811/Library/Application Support/typora-user-images/image-20210701102351548.png)

  ï¼ˆä¸è¿‡å‰æå¿…é¡»æ˜¯adapteræ·»åŠ åœ¨3â€˜ç«¯ï¼‰

  è¿™é‡Œæä¸€ä¸‹RabbitQCçš„adapterä¿®å‰ªï¼šå¯¹äºå•ç«¯æ•°æ®ï¼Œå¼€å§‹å…ˆæ‹¿å‰100wæ¡readæ£€æµ‹adapterï¼Œç„¶åæ¯æ¬¡ç›´æ¥ä¿®å‰ªï¼›å¯¹äºåŒç«¯çš„æ•°æ®ï¼Œè¿‡ç¨‹å’ŒafterQCç±»ä¼¼ï¼Œå…ˆè¿›è¡ŒOverlapAnalysisï¼ˆå¦‚æœæ‰“å¼€äº†correctå‚æ•°ä¼šæ¥ç€è¿›è¡Œä¿®æ­£ï¼ŒåŒ…å«äº†Sequencing error profilingéƒ¨åˆ†ï¼Œé»˜è®¤æ˜¯ä¸å¼€çš„ï¼‰ï¼Œå¦‚æœoffest<0ï¼ŒæŒ‰ç…§ğŸ‘†çš„æ–¹æ³•å»é™¤adapterï¼Œå¦åˆ™ï¼Œå’Œå•ç«¯ä¸€æ ·ï¼ˆä½†æ˜¯åŒç«¯æ•°æ®é»˜è®¤ä¸€å¼€å§‹ä¸åšadapterçš„æ£€æµ‹ï¼‰

- Quality profilingï¼šstrand bias profiling to reflect amplification bias, and per-cycle dis- continuity profiling to reflect sequencing quality instability. å‰è€…æ˜¯ç»Ÿè®¡çŸ­kmeræ­£åå‘è®¡æ•°æ˜¯ä¸æ˜¯åŸºæœ¬ç›¸åŒï¼Œåè€…ç»Ÿè®¡ä¸è¿ç»­æ€§æ˜¯ä¸æ˜¯ç›¸å¯¹ç¨³å®šã€‚

- VS

  ![image-20210701110116765](/Users/ylf9811/Library/Application Support/typora-user-images/image-20210701110116765.png)

## fastp

- Adapter trimmingï¼šSE-adapter sequences are detected by assembling the high-frequency read tails//TODOï¼›PE-adapt- er sequences are detected by finding the overlap of each pair.

  The adapter-sequence detection algorithm is based on two assumptions: the first is that only one adapter exists in the data; the second is that adapter sequences exist only in the read tails.

- Base correctionï¼šğŸ‘Œ

- Sliding window quality pruningï¼šå’ŒğŸ‘†Automatic trimmingç±»ä¼¼ï¼ŒåŠ ä¸Šäº†æ»‘åŠ¨çª—å£çš„æ€æƒ³

- polyG and polyX tail trimmingï¼š//TODOå…·ä½“å’‹åš

- UMI preprocessingï¼š//TODO a sample index or inserted DNAï¼Ÿè®ºæ–‡é‡Œå¤§æ¦‚è¯´äº†fastpæŠŠå·²æœ‰çš„umiå·¥å…·çš„åŠŸèƒ½åˆæˆè¿›æ¥äº†ï¼Œä½†æ˜¯å¹¶æ²¡æœ‰è¯¦ç»†è¯´æ˜åŠŸèƒ½æ˜¯ä»€ä¹ˆã€‚

- Output splittingï¼šsplitting by file lines and splitting by file numbers//TODO

- Duplication evaluationï¼š

- Overrepresented sequence analysisï¼š



## MultiQC

a tool to create a single report visualising output from multiple tools across many samples, enabling global trends and biases to be quickly identified.

//TODO confounding batch effects ? 

- Leek,J.T. et al. (2010) Tackling the widespread and critical impact of batch ef- fects in high-throughput data. Nat. Rev. Genet., 11, 733739.
- Meyer,C.A. and Liu,X.S. (2014) Identifying and mitigating bias in next- generation sequencing methods for chromatin biology. Nat. Rev. Genet., 15, 709721.
- Taub,M.A. et al. (2010) Overcoming bias and systematic errors in next gener- ation sequencing data. Genome Med., 2, 87

## FQC

ğŸ‘†distinguishes FQC from similar tools (e.g. MultiQC; Ewels et al., 2016) designed to summarize FASTQ data sets as individual or groups of samples, but that lack the ability to display multiple, single-sample reports in a unified dashboard.

## fastqc

https://www.bioinformatics.babraham.ac.uk/projects/fastqc/Help/3%20Analysis%20Modules/

- Basic Statistics

- Per Base Sequence Quality

- Per Sequence Quality Scores

- Per Base Sequence Content

- Per Sequence GC Content

- Per Base N Content

- Sequence Length Distribution

- Duplicate Sequences

- Overrepresented Sequences

- Adapter Content

- Kmer Content

- Per Tile Sequence Quality



Parser 

io -> zip  | æµå¼å¤„ç† ï½œ

thread ->  threadPool ï½œ 

functions

Vi 

## 0706

now cmd parser use CLI11, IO use RabbitIO, 



#### 1.æ–‡ä»¶å‘½åè§„åˆ™

æ–‡ä»¶åå…¨éƒ¨å°å†™ï¼Œå¯ä»¥å«ä¸‹åˆ’çº¿æˆ–è¿å­—ç¬¦ï¼ŒæŒ‰é¡¹ç›®çº¦å®šå‘½å,ä¸”å°½é‡ä¿è¯æ–‡ä»¶åæ˜ç¡®ã€‚æ¯”å¦‚ï¼š
`cmd_save_player_info_class.cc ï¼Œmy_use_full_class.cc`

å®šä¹‰ç±»çš„æ–‡ä»¶åä¸€èˆ¬æ˜¯æˆå¯¹å‡ºç°ï¼Œå¦‚ï¼š`foo_bar.h foo_bar.cc`

è‹¥æ˜¯ç±»ä¸­å«å¤§é‡å†…è”å‡½æ•°ï¼Œæˆ‘ä»¬è¿˜å¯ä½¿ç”¨-ini.hæ–‡ä»¶ï¼Œä½¿ä¹‹æ–‡ä»¶å†…å®¹æ›´åŠ æ¸…æ™°ï¼Œäºæ˜¯åˆå¦‚ï¼š
`url_table.h url_table.cc url-table-ini.h`

#### 2.ç±»å‘½åè§„åˆ™

ç±»å‹å‘½åæ¯ä¸ªå•è¯é¦–å­—æ¯å¤§å†™ï¼Œä¸å«ä¸‹åˆ’çº¿ï¼Œä»¥åè¯å½¢å¼ã€‚æ¯”å¦‚ï¼š`MyPalyerManager`ï¼Œè¿™å¯¹äºæ‰€æœ‰ç±»å‹å‘½åä¸€æ ·ï¼Œç±»ç»“æ„ä½“ï¼Œæšä¸¾ï¼Œç±»å®šä¹‰éƒ½æ˜¯å¦‚æ­¤ï¼Œæ¯”å¦‚ï¼š`MyExcitingEnum`

#### 3.å˜é‡å‘½åè§„åˆ™

å˜é‡åä¸€å¾‹å°å†™ï¼Œå•è¯ç”¨ä¸‹åˆ’çº¿ç›¸è¿ï¼Œä¾‹å¦‚ï¼š`int player_id; string table_name;`

ç‰¹æ®Šçš„æ˜¯ç±»æˆå‘˜å˜é‡ï¼Œåè·Ÿä¸‹åˆ’çº¿åŒºåˆ«æ™®é€šå˜é‡ï¼Œæ¯”å¦‚ï¼š`player_name_ player_id_`

å…¨å±€å˜é‡åˆ™ä»¥ g_ å¼€å¤´ï¼Œæ¯”å¦‚ ï¼š`g_system_time`

å½“ç„¶ç»“æ„ä½“æˆå‘˜å˜é‡è¿˜æ˜¯å’Œæ™®é€šå˜é‡ä¸€æ ·,æ¯”å¦‚ï¼š`string name; int num_entries;`

#### 4.å¸¸é‡å‘½åè§„åˆ™

kåé¢è·Ÿå¤§å†™å­—æ¯å¼€å¤´çš„å•è¯ï¼Œæ¯”å¦‚ï¼š
`const int kDaysInAWeek=7; const string kCompanyName=â€Tecentâ€;`

#### 5.å‡½æ•°å‘½åè§„åˆ™

å¸¸è§„å‡½æ•°æ¯ä¸ªå•è¯é¦–å­—æ¯å¤§å†™ï¼Œä½¿ç”¨å‘½ä»¤å¼è¯­æ°”ï¼Œæ¯”å¦‚ï¼š`OpenFile() CheckFileName()`ï¼Œ

è€Œå­˜å–å‡½æ•°æˆ–çŸ­å°çš„å†…è”å‡½æ•°ä½¿ç”¨å°å†™åŠ ä¸‹åˆ’çº¿ï¼Œä¸”ä¸è®¿é—®å˜é‡ç›¸å»åˆï¼Œæ¯”å¦‚`set_num_errors();`

```cpp
class Player{ 
public: 
void set_player_id(const int player_id){return player_id_=player_id;} 
int get_player_id() const{return player-id_;} 
private: 
int palyer_id_; 
};
```

#### 7.æšä¸¾å‘½åè§„åˆ™

æšä¸¾ç±»åå±äºç±»å‹åï¼ŒæŒ‰ç±»å‘½åï¼Œæšä¸¾å€¼å…¨å¤§å†™åŠ ä¸‹åˆ’çº¿ï¼Œæ¯”å¦‚ï¼š`ENUM_NAME` ã€‚

#### 8.å®å˜é‡å‘½åè§„åˆ™

å¦‚æœä½ ä¸€å®šè¦ç”¨åˆ°å®ï¼Œå…¨å¤§å†™åŠ ä¸‹åˆ’çº¿ï¼Œæ¯”å¦‚ï¼š`define PI_ROUND 3.0`ã€‚

#### 9.æ ¼å¼ç¾åŒ–

å¯ä»¥å€ŸåŠ©å·¥å…·è¿›è¡Œç¾åŒ–ã€‚æ–¹ä¾¿å¿«æ·ã€‚æ¯”å¦‚è¯´æˆ‘ç”¨çš„Qté‡Œé¢çš„Beautifierï¼Œå°±å¯ä»¥è¿›è¡Œä¸€é”®æ ¼å¼åŒ–ä»£ç ã€‚

https://www.jianshu.com/p/f56383486520

#### 8.includeè§„èŒƒ

cåº“ã€c++åº“ã€å…¶ä»–åº“ã€æœ¬åœ°æ–‡ä»¶





## 0707



æ ¹æ®é‚£å¤©å’Œå¤§å¸ˆå…„çš„è®¨è®ºç»“æœï¼ŒåŸºæœ¬ç¡®å®šæ¯ä¸ªåŠŸèƒ½ä¸€ä¸ªç±»ï¼Œæš‚æ—¶è¿™å‡ è¿™å‡ ä¸ªç±»ï¼š

- Filter
- Trimerï¼šåŒ…å«front tail trimå’Œadapter trim
- Umi
- PolyX
- Overrepresented
- Duplicate

| Version              | Se    | Pe   |
| -------------------- | ----- | ---- |
| count lines          | 5.60  |      |
| count bases thread 1 | 12.66 |      |
| count bases thread 4 | 3.42  |      |



ğŸ‘†æ˜¯ç®€å•çš„seæ•°æ®çš„ç®€å•ä¿¡æ¯ç»Ÿè®¡ï¼ŒğŸ‘‡åŠ ä¸€ç‚¹fliterï¼Œæ–¹ä¾¿åšè¾“å‡ºã€‚

afterQCä¸­æ˜¯è¿™æ ·åšçš„ï¼š

```
quality filters->count the number of low quality bases or N, calculate the mean quality of each read, and then determine whether to keep or discard this read.
```

fastpä¸­æ˜¯ï¼š

```
static const int PASS_FILTER = 0;
static const int FAIL_POLY_X = 4;
static const int FAIL_OVERLAP = 8;
static const int FAIL_N_BASE = 12;
static const int FAIL_LENGTH = 16;
static const int FAIL_TOO_LONG = 17;
static const int FAIL_QUALITY = 20;
static const int FAIL_COMPLEXITY = 24;
```

è¿™é‡Œå…ˆå®ç°å…¶ä¸­çš„ 0 12 16 17 20

|                            | Se    |      |
| -------------------------- | ----- | ---- |
| add sample filter thread 1 | 13.69 |      |
| add sample filter thread 4 | 3.67  |      |
|                            |       |      |

æœ‰äº†ç®€å•çš„è¿‡æ»¤ä¹‹åå°±æœ‰è¾“å‡ºè¿‡æ»¤åreadçš„å¿…è¦çš„ï¼Œä¸‹é¢å…ˆç®€å•å®ç°ä¸€ç‰ˆoutputã€‚

æ‰æƒ³neoReferenceçš„ç©ºé—´ä½¿ç”¨ä¸ä¼šå¤ªå¤šï¼Œæ‰€ä»¥ç›´æ¥å­˜åº”è¯¥å†…å­˜ä¹Ÿæ˜¯å¤Ÿçš„ï¼Œä¸ç”¨åˆ†æ‰¹å¤„ç†ã€‚

è¿™ä¸€ç‰ˆå†™çš„ä¼¼ä¹å¹¶ä¸å·§å¦™ï¼Œåœ¨ç»Ÿè®¡ä¿¡æ¯å’Œè¿‡æ»¤çš„åŒæ—¶å¯¹pass_dataè¿›è¡Œæ‹·è´ï¼Œæ‹·è´åˆ°è¿ç»­çš„å†…å­˜ä¸­ï¼Œæ¯64Måšæˆä¸€ä¸ªstringï¼Œç„¶åç”¨æ— é”é˜Ÿåˆ—ç»´æŠ¤ï¼Œä¸æ­¤åŒæ—¶å¼€ä¸€ä¸ªå†™çº¿ç¨‹æ£€æµ‹é˜Ÿåˆ—æ˜¯å¦ä¸ºç©ºå¹¶è¿›è¡Œè¾“å‡ºã€‚

|                                                              | Se    |      |
| ------------------------------------------------------------ | ----- | ---- |
| add simple output thread 1ï¼ˆconcurrentqueue.hï¼‰              | 38.42 |      |
| add simple output thread 4ï¼ˆconcurrentqueue.hï¼‰              | 11.94 |      |
| adjust output block size and optimize queue(reserve) thread 1 | 36.74 |      |
| adjust output block size and optimize queue(reserve) thread 4 | 10.84 |      |
| adjust output block size and optimize queue(reserve) thread 1 just no write | 36.79 |      |
| adjust output block size and optimize queue(reserve) thread 4 just no write | 9.95  |      |

ç°åœ¨å•çº¿ç¨‹æ…¢æ˜¯ä¸€æ¬¡å¤šä½™çš„æ‹·è´ï¼Œå¤šçº¿ç¨‹åŠ é€Ÿæ¯”ä¸€èˆ¬å¤§æ¦‚ç‡æ˜¯å› ä¸ºæ— é”é˜Ÿåˆ—ï¼Œå¯ä»¥è€ƒè™‘æ¢æˆåŸå­æ“ä½œã€‚

ğŸ‘†è€ƒè™‘åˆ°é˜Ÿåˆ—æ“ä½œå¹¶ä¸å¤š(fileSize/4M)ï¼Œé—®é¢˜ä¸å¤§ï¼Œé‡ç‚¹è¿˜æ˜¯ä¼˜åŒ–é‚£ä¸€æ¬¡æ‹·è´ã€‚

|                                             | Se    |      |
| ------------------------------------------- | ----- | ---- |
| One less memory copy thread 1               | 15.13 |      |
| One less memory copy thread 4               | 15.33 |      |
| One less memory copy thread 1 just no write | 15.03 |      |
| One less memory copy thread 4 just no write | 3.85  |      |
|                                             |       |      |

åŸºæœ¬ç¬¦åˆé¢„æœŸï¼Œå‡å°‘æ‹·è´ä¹‹åå¿«äº†ä¸€å€å·¦å³ï¼Œä½†æ˜¯å¤šçº¿ç¨‹çš„æ—¶å€™å¡åœ¨å†™çš„è¿‡ç¨‹ï¼ŒæŠŠwriteæ³¨é‡Šå°±åŠ é€Ÿæ¯”å¾ˆå¥½äº†ã€‚